pkg/pipeline/conntrack/hash.go

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2022 IBM, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package conntrack
+
+import (
+	"bytes"
+	"encoding/gob"
+	"fmt"
+	"hash/fnv"
+
+	"github.com/netobserv/flowlogs-pipeline/pkg/api"
+	"github.com/netobserv/flowlogs-pipeline/pkg/config"
+)
+
+// ComputeHash computes the hash of a flow log according to keyFields.
+// Two flow logs will have the same hash if they belong to the same connection.
+func ComputeHash(flowLog config.GenericMap, keyFields api.KeyFields) ([]byte, error) {
+	type hashType []byte
+	fieldGroup2hash := make(map[string]hashType)
+
+	// Compute the hash of each field group
+	for _, fg := range keyFields.FieldGroups {
+		h, err := computeHashFields(flowLog, fg.Fields)
+		if err != nil {
+			return nil, fmt.Errorf("compute hash: %w", err)
+		}
+		fieldGroup2hash[fg.Name] = h
+	}
+
+	// Compute the total hash
+	hash := fnv.New32a()
+	for _, fgName := range keyFields.Hash.FieldGroupRefs {
+		hash.Write(fieldGroup2hash[fgName])
+	}
+	if keyFields.Hash.FieldGroupARef != "" {
+		hashA := fieldGroup2hash[keyFields.Hash.FieldGroupARef]
+		hashB := fieldGroup2hash[keyFields.Hash.FieldGroupBRef]
+		// Determine order between A's and B's hash to get the same hash for both flow logs from A to B and from B to A.
+		if bytes.Compare(hashA, hashB) < 0 {
+			hash.Write(hashA)
+			hash.Write(hashB)
+		} else {
+			hash.Write(hashB)
+			hash.Write(hashA)
+		}
+	}
+	return hash.Sum([]byte{}), nil
+}
+
+func computeHashFields(flowLog config.GenericMap, fieldNames []string) ([]byte, error) {
+	h := fnv.New32a()
+	for _, fn := range fieldNames {
+		// TODO: How should we handle a missing fieldName?
```

Comment:
Info log and skip ( for now) --- this will cause in different Hash anyway ... maybe later we will make that a parameter